In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/players-all/players.xls
/kaggle/input/baseball-databank/TeamsHalf.csv
/kaggle/input/baseball-databank/SeriesPost.csv
/kaggle/input/baseball-databank/AwardsSharePlayers.csv
/kaggle/input/baseball-databank/Batting.csv
/kaggle/input/baseball-databank/AwardsShareManagers.csv
/kaggle/input/baseball-databank/Salaries.csv
/kaggle/input/baseball-databank/Teams.csv
/kaggle/input/baseball-databank/Fielding.csv
/kaggle/input/baseball-databank/Pitching.csv
/kaggle/input/baseball-databank/AwardsManagers.csv
/kaggle/input/baseball-databank/ManagersHalf.csv
/kaggle/input/baseball-databank/BattingPost.csv
/kaggle/input/baseball-databank/TeamsFranchises.csv
/kaggle/input/baseball-databank/AwardsPlayers.csv
/kaggle/input/baseball-databank/Master.csv
/kaggle/input/baseball-databank/AllstarFull.csv
/kaggle/input/baseball-databank/HallOfFame.csv
/kaggle/input/baseball-databank/FieldingOF.csv
/kaggle/input/baseball-databank/PitchingPost.csv
/kaggle/input/baseball-databank/Managers.csv
/kaggle/

In [2]:
batting=pd.read_csv('/kaggle/input/career-hitting/batting.csv').drop(columns='Unnamed: 0')
batting

player_id  stint       g       ab       r       h  double  triple  \
0     mcguide01   37.0  1781.0   6290.0   770.0  1748.0   300.0    79.0   
1     newsobo01   39.0   609.0   1337.0    86.0   253.0    28.0     3.0   
2     henderi01   33.0  3081.0  10961.0  2295.0  3055.0   510.0    66.0   
3      kaatji01   31.0  1004.0   1251.0   117.0   232.0    44.0     5.0   
4      johnto01   30.0   761.0    900.0    43.0   141.0    13.0     0.0   
...         ...    ...     ...      ...     ...     ...     ...     ...   
3798  pasquda01   10.0   905.0   2620.0   341.0   638.0   129.0    15.0   
3799   fainfe01   11.0  1151.0   3930.0   595.0  1139.0   213.0    30.0   
3800  reynoma01   11.0  1258.0   4178.0   612.0   959.0   192.0    13.0   
3801  madsory01   10.0   559.0     48.0     3.0     6.0     1.0     0.0   
3802  brognri01   11.0   848.0   2958.0   379.0   795.0   176.0    13.0   

         hr     rbi      sb     cs      bb      so   ibb   hbp    sh    sf  \
0      45.0   840.0   117.0    0.0   515.0   214.0   0.0  84.0  43.0   0.0   
1       1.0   102.0     1.0    1.0    17.0   266.0   0.0   1.0  86.0   0.0   
2     297.0  1115.0  1406.0  335.0  2190.0  1694.0  61.0  98.0  30.0  67.0   
3      16.0   106.0     5.0    1.0    63.0   367.0   1.0   7.0  34.0   7.0   
4       5.0    54.0     0.0    0.0    33.0   211.0   0.0   1.0  93.0   3.0   
...     ...     ...     ...    ...     ...     ...   ...   ...   ...   ...   
3798  117.0   390.0     7.0   10.0   335.0   642.0  29.0  15.0   9.0  21.0   
3799   48.0   570.0    46.0   28.0   904.0   261.0   3.0  18.0  44.0   8.0   
3800  237.0   661.0    59.0   28.0   550.0  1519.0  24.0  47.0   3.0  34.0   
3801    0.0     2.0     0.0    0.0     2.0    20.0   0.0   0.0   7.0   0.0   
3802  106.0   458.0    32.0   16.0   227.0   655.0  29.0   7.0   5.0  26.0   

      g_idp  
0       0.0  
1      30.0  
2     172.0  
3      14.0  
4      20.0  
...     ...  
3798   43.0  
3799  117.0  
3800   88.0  
3801    2.0  
3802   72.0  

[3803 rows x 19 columns]

In [3]:
master=pd.read_excel('/kaggle/input/correct/correct.xlsx')
master=master[['IDPLAYER','PLAYERNAME']]
master

IDPLAYER     PLAYERNAME
0     aardsda01  David Aardsma
1      abadfe01  Fernando Abad
2     abreual01   Albert Abreu
3     abreubo01    Bobby Abreu
4     abreubr01    Bryan Abreu
...         ...            ...
2505   zitoba01     Barry Zito
2506  zobribe01    Ben Zobrist
2507  zuninmi01    Mike Zunino
2508   zychto01      Tony Zych
2509  zzzzzzz01    Last Player

[2510 rows x 2 columns]

In [4]:
master['IDPLAYER'].str.contains('abreubo01').any()

True

In [5]:
players=[]
for player in batting['player_id']:
    if master['IDPLAYER'].str.contains(player).any():
        players.append(player)

In [6]:
len(players)

410

In [7]:
len(batting)

3803

Dataset of player names is inadequate because it only accounts for 410 of 3803 player IDs we need.

In [8]:
hof=pd.read_csv('/kaggle/input/baseball-databank/HallOfFame.csv')
hof

playerID  yearid votedBy  ballots  needed  votes inducted category  \
0      cobbty01    1936   BBWAA    226.0   170.0  222.0        Y   Player   
1      ruthba01    1936   BBWAA    226.0   170.0  215.0        Y   Player   
2     wagneho01    1936   BBWAA    226.0   170.0  215.0        Y   Player   
3     mathech01    1936   BBWAA    226.0   170.0  205.0        Y   Player   
4     johnswa01    1936   BBWAA    226.0   170.0  189.0        Y   Player   
...         ...     ...     ...      ...     ...    ...      ...      ...   
4115  grudzma01    2016   BBWAA    440.0   330.0    0.0        N   Player   
4116  ausmubr01    2016   BBWAA    440.0   330.0    0.0        N   Player   
4117  hamptmi01    2016   BBWAA    440.0   330.0    0.0        N   Player   
4118  castilu01    2016   BBWAA    440.0   330.0    0.0        N   Player   
4119  lowelmi01    2016   BBWAA    440.0   330.0    0.0        N   Player   

     needed_note  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
...          ...  
4115         NaN  
4116         NaN  
4117         NaN  
4118         NaN  
4119         NaN  

[4120 rows x 9 columns]

HOF dataset has a little more than the batting dataset- this is probably due to pitchers or managers.

In [9]:
#df of all players included in in HOF dataset
ballot=batting[batting['player_id'].isin(hof['playerID'].tolist())]
ballot

player_id  stint       g       ab       r       h  double  triple  \
1     newsobo01   39.0   609.0   1337.0    86.0   253.0    28.0     3.0   
2     henderi01   33.0  3081.0  10961.0  2295.0  3055.0   510.0    66.0   
3      kaatji01   31.0  1004.0   1251.0   117.0   232.0    44.0     5.0   
4      johnto01   30.0   761.0    900.0    43.0   141.0    13.0     0.0   
5     carltst01   31.0   745.0   1719.0   123.0   346.0    49.0     6.0   
...         ...    ...     ...      ...     ...     ...     ...     ...   
3729  robinja02   10.0  1382.0   4877.0   947.0  1518.0   273.0    54.0   
3738  torboje01   10.0   574.0   1391.0    78.0   297.0    42.0     3.0   
3745  hornebo01   10.0  1020.0   3777.0   560.0  1047.0   169.0     8.0   
3774  tobinji01   11.0   396.0    796.0    81.0   183.0    35.0     3.0   
3782  kremera01   10.0   309.0    687.0    57.0   122.0    30.0     2.0   

         hr     rbi      sb     cs      bb      so   ibb   hbp     sh    sf  \
1       1.0   102.0     1.0    1.0    17.0   266.0   0.0   1.0   86.0   0.0   
2     297.0  1115.0  1406.0  335.0  2190.0  1694.0  61.0  98.0   30.0  67.0   
3      16.0   106.0     5.0    1.0    63.0   367.0   1.0   7.0   34.0   7.0   
4       5.0    54.0     0.0    0.0    33.0   211.0   0.0   1.0   93.0   3.0   
5      13.0   140.0     2.0    2.0    41.0   413.0   0.0  10.0   94.0  13.0   
...     ...     ...     ...    ...     ...     ...   ...   ...    ...   ...   
3729  137.0   734.0   197.0   30.0   740.0   291.0   7.0  72.0  104.0   9.0   
3738    8.0   101.0     3.0    6.0   103.0   189.0  29.0   4.0   17.0  10.0   
3745  218.0   685.0    14.0   18.0   369.0   512.0  39.0  16.0    1.0  50.0   
3774   17.0   102.0     1.0    0.0    80.0   162.0   0.0   3.0    5.0   0.0   
3782    5.0    69.0     0.0    0.0    37.0   213.0   0.0   0.0   36.0   0.0   

      g_idp  
1      30.0  
2     172.0  
3      14.0  
4      20.0  
5      33.0  
...     ...  
3729  113.0  
3738   36.0  
3745  111.0  
3774   22.0  
3782    0.0  

[1127 rows x 19 columns]

In [10]:
#df of all players not in HOF dataset
absent=batting[-batting['player_id'].isin(hof['playerID'].tolist())]
absent

player_id  stint       g      ab       r       h  double  triple     hr  \
0     mcguide01   37.0  1781.0  6290.0   770.0  1748.0   300.0    79.0   45.0   
6     moyerja01   29.0   701.0   400.0    24.0    51.0     5.0     0.0    0.0   
14    weathda01   34.0   967.0   139.0     7.0    14.0     0.0     0.0    2.0   
17    francju01   27.0  2527.0  8677.0  1285.0  2586.0   407.0    54.0  173.0   
21    morgami01   28.0   597.0   497.0    13.0    54.0     3.0     1.0    0.0   
...         ...    ...     ...     ...     ...     ...     ...     ...    ...   
3798  pasquda01   10.0   905.0  2620.0   341.0   638.0   129.0    15.0  117.0   
3799   fainfe01   11.0  1151.0  3930.0   595.0  1139.0   213.0    30.0   48.0   
3800  reynoma01   11.0  1258.0  4178.0   612.0   959.0   192.0    13.0  237.0   
3801  madsory01   10.0   559.0    48.0     3.0     6.0     1.0     0.0    0.0   
3802  brognri01   11.0   848.0  2958.0   379.0   795.0   176.0    13.0  106.0   

         rbi     sb     cs     bb      so   ibb   hbp    sh    sf  g_idp  
0      840.0  117.0    0.0  515.0   214.0   0.0  84.0  43.0   0.0    0.0  
6       15.0    0.0    0.0   35.0   152.0   0.0   1.0  64.0   2.0    5.0  
14       4.0    0.0    0.0    7.0    85.0   0.0   1.0  17.0   0.0    0.0  
17    1194.0  281.0  107.0  917.0  1341.0  67.0  38.0  19.0  80.0  312.0  
21      15.0    0.0    0.0   13.0   151.0   0.0   1.0  59.0   3.0    5.0  
...      ...    ...    ...    ...     ...   ...   ...   ...   ...    ...  
3798   390.0    7.0   10.0  335.0   642.0  29.0  15.0   9.0  21.0   43.0  
3799   570.0   46.0   28.0  904.0   261.0   3.0  18.0  44.0   8.0  117.0  
3800   661.0   59.0   28.0  550.0  1519.0  24.0  47.0   3.0  34.0   88.0  
3801     2.0    0.0    0.0    2.0    20.0   0.0   0.0   7.0   0.0    2.0  
3802   458.0   32.0   16.0  227.0   655.0  29.0   7.0   5.0  26.0   72.0  

[2676 rows x 19 columns]

In [11]:
absent.sort_values(by='rbi',ascending=False)

player_id  stint       g       ab       r       h  double  triple  \
150   rodrial01   21.0  2719.0  10341.0  2002.0  3070.0   541.0    31.0   
174   ramirma02   23.0  2302.0   8244.0  1544.0  2574.0   547.0    20.0   
29    thomeji01   28.0  2543.0   8422.0  1583.0  2328.0   451.0    26.0   
1274  pujolal01   15.0  2274.0   8545.0  1599.0  2666.0   583.0    16.0   
395   ortizda01   19.0  2257.0   8103.0  1340.0  2303.0   584.0    18.0   
...         ...    ...     ...      ...     ...     ...     ...     ...   
3455  norrimi01   10.0   204.0      1.0     1.0     0.0     0.0     0.0   
359   anderla02   21.0   699.0     38.0     4.0     5.0     0.0     0.0   
1724   nagych01   14.0   320.0     19.0     3.0     2.0     0.0     0.0   
3162   reedje01   13.0   239.0      0.0     0.0     0.0     0.0     0.0   
3294  havenbr01   12.0   205.0      3.0     0.0     0.0     0.0     0.0   

         hr     rbi     sb    cs      bb      so    ibb    hbp    sh     sf  \
150   687.0  2055.0  326.0  76.0  1324.0  2220.0   97.0  175.0  16.0  108.0   
174   555.0  1831.0   38.0  33.0  1329.0  1813.0  216.0  109.0   2.0   90.0   
29    612.0  1699.0   19.0  20.0  1747.0  2548.0  173.0   69.0   1.0   74.0   
1274  560.0  1698.0  103.0  41.0  1165.0   978.0  296.0   98.0   1.0   93.0   
395   503.0  1641.0   15.0   9.0  1239.0  1664.0  194.0   36.0   2.0   85.0   
...     ...     ...    ...   ...     ...     ...    ...    ...   ...    ...   
3455    0.0     0.0    0.0   0.0     0.0     1.0    0.0    0.0   0.0    0.0   
359     0.0     0.0    0.0   0.0     3.0    15.0    0.0    0.0   4.0    0.0   
1724    0.0     0.0    0.0   0.0     0.0    10.0    0.0    0.0   1.0    0.0   
3162    0.0     0.0    0.0   0.0     0.0     0.0    0.0    0.0   0.0    0.0   
3294    0.0     0.0    0.0   0.0     0.0     1.0    0.0    0.0   0.0    0.0   

      g_idp  
150   257.0  
174   243.0  
29    165.0  
1274  312.0  
395   214.0  
...     ...  
3455    0.0  
359     0.0  
1724    0.0  
3162    0.0  
3294    0.0  

[2676 rows x 19 columns]

Lot of these players are current players (dataset is updated as of 2016), and many would never get votes for the hall. For example, let's see how many of these players have at least 2000 hits.

In [12]:
hits=absent[absent['h']>=2000]
hits

player_id  stint       g       ab       r       h  double  triple  \
17    francju01   27.0  2527.0   8677.0  1285.0  2586.0   407.0    54.0   
29    thomeji01   28.0  2543.0   8422.0  1583.0  2328.0   451.0    26.0   
45    vizquom01   24.0  2968.0  10586.0  1445.0  2877.0   456.0    77.0   
74    rodriiv01   25.0  2543.0   9592.0  1354.0  2844.0   572.0    51.0   
137   giambja01   22.0  2260.0   7267.0  1227.0  2010.0   405.0     9.0   
150   rodrial01   21.0  2719.0  10341.0  2002.0  3070.0   541.0    31.0   
156   hinespa01   22.0  1659.0   7062.0  1217.0  2134.0   399.0    93.0   
174   ramirma02   23.0  2302.0   8244.0  1544.0  2574.0   547.0    20.0   
205   abreubo01   22.0  2425.0   8480.0  1453.0  2470.0   574.0    59.0   
213   donovpa01   23.0  1821.0   7496.0  1318.0  2253.0   207.0    75.0   
243   ramirar01   22.0  2194.0   8136.0  1098.0  2303.0   495.0    24.0   
249   beltrca01   22.0  2306.0   8749.0  1449.0  2454.0   503.0    78.0   
250   jeterde01   20.0  2747.0  11195.0  1923.0  3465.0   544.0    66.0   
257   ibanera01   21.0  2161.0   7471.0  1055.0  2034.0   424.0    51.0   
312   jonesch06   19.0  2499.0   8984.0  1619.0  2726.0   549.0    38.0   
318   hunteto01   19.0  2372.0   8857.0  1296.0  2452.0   498.0    39.0   
321   daviswi02   20.0  2429.0   9174.0  1217.0  2561.0   395.0   138.0   
353   konerpa01   20.0  2349.0   8393.0  1162.0  2340.0   410.0     8.0   
370   burnsge02   22.0  1866.0   6573.0   901.0  2018.0   444.0    72.0   
386   rolensc01   21.0  2038.0   7398.0  1211.0  2077.0   517.0    43.0   
395   ortizda01   19.0  2257.0   8103.0  1340.0  2303.0   584.0    18.0   
474   polanpl01   20.0  1927.0   7214.0  1009.0  2142.0   348.0    32.0   
488    ryanji01   18.0  2012.0   8164.0  1642.0  2502.0   451.0   157.0   
506   cabreor01   21.0  1985.0   7562.0   985.0  2055.0   459.0    32.0   
510   corcoto01   18.0  2200.0   8804.0  1184.0  2252.0   289.0   155.0   
578   beltrad01   18.0  2567.0   9712.0  1339.0  2767.0   560.0    35.0   
582   damonjo01   18.0  2490.0   9736.0  1668.0  2769.0   522.0   109.0   
599   tejadmi01   18.0  2171.0   8434.0  1230.0  2407.0   468.0    23.0   
658   soriaal01   18.0  1975.0   7750.0  1152.0  2095.0   481.0    31.0   
771   heltoto01   17.0  2247.0   7962.0  1401.0  2519.0   592.0    37.0   
902   renteed01   16.0  2152.0   8142.0  1200.0  2327.0   436.0    29.0   
912   suzukic01   17.0  2357.0   9362.0  1348.0  2935.0   341.0    91.0   
924     leeca01   18.0  2099.0   7983.0  1125.0  2273.0   469.0    19.0   
947   koneted01   17.0  2085.0   7649.0   972.0  2150.0   344.0   182.0   
981   guerrvl01   16.0  2147.0   8155.0  1328.0  2590.0   477.0    46.0   
1057  rolliji01   16.0  2234.0   9145.0  1396.0  2422.0   503.0   114.0   
1111  ordonma01   15.0  1848.0   6978.0  1076.0  2156.0   426.0    21.0   
1161  youngmi02   16.0  1970.0   7918.0  1137.0  2375.0   441.0    60.0   
1274  pujolal01   15.0  2274.0   8545.0  1599.0  2666.0   583.0    16.0   
1513    hoydu01   14.0  1796.0   7112.0  1426.0  2044.0   248.0   121.0   
1576  pierrju01   14.0  1994.0   7525.0  1075.0  2217.0   255.0    94.0   
2048  cabremi01   13.0  1938.0   7258.0  1229.0  2331.0   492.0    16.0   
2185  mckeaed01   13.0  1654.0   6890.0  1227.0  2083.0   272.0   158.0   
3009   canoro01   11.0  1687.0   6555.0   958.0  2015.0   446.0    31.0   

         hr     rbi     sb     cs      bb      so    ibb    hbp     sh     sf  \
17    173.0  1194.0  281.0  107.0   917.0  1341.0   67.0   38.0   19.0   80.0   
29    612.0  1699.0   19.0   20.0  1747.0  2548.0  173.0   69.0    1.0   74.0   
45     80.0   951.0  404.0  167.0  1028.0  1087.0   25.0   49.0  256.0   94.0   
74    311.0  1332.0  127.0   64.0   513.0  1474.0   67.0   58.0   31.0   76.0   
137   440.0  1441.0   20.0   12.0  1366.0  1572.0   97.0  180.0    2.0   93.0   
150   687.0  2055.0  326.0   76.0  1324.0  2220.0   97.0  175.0   16.0  108.0   
156    57.0   855.0  163.0   11.0   372.0   309.0    0.0   36.0

In [13]:
len(hits)

44

Only 44 have 2000 hits, and I recognize pretty much all of the players because they're either still playing or were not yet eligible for the hall of fame as of 2016 (have to be retired for at least five years).

In [14]:
hof[hof['playerID'].isin(batting['player_id'].tolist())]

playerID  yearid votedBy  ballots  needed  votes inducted category  \
0      cobbty01    1936   BBWAA    226.0   170.0  222.0        Y   Player   
1      ruthba01    1936   BBWAA    226.0   170.0  215.0        Y   Player   
2     wagneho01    1936   BBWAA    226.0   170.0  215.0        Y   Player   
3     mathech01    1936   BBWAA    226.0   170.0  205.0        Y   Player   
4     johnswa01    1936   BBWAA    226.0   170.0  189.0        Y   Player   
...         ...     ...     ...      ...     ...    ...      ...      ...   
4115  grudzma01    2016   BBWAA    440.0   330.0    0.0        N   Player   
4116  ausmubr01    2016   BBWAA    440.0   330.0    0.0        N   Player   
4117  hamptmi01    2016   BBWAA    440.0   330.0    0.0        N   Player   
4118  castilu01    2016   BBWAA    440.0   330.0    0.0        N   Player   
4119  lowelmi01    2016   BBWAA    440.0   330.0    0.0        N   Player   

     needed_note  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
...          ...  
4115         NaN  
4116         NaN  
4117         NaN  
4118         NaN  
4119         NaN  

[3964 rows x 9 columns]

Problem with appending the HOF dataset is that players will show up more than once for each time they show up on the ballot. It took Craig Biggio years to finally reach the thershold of votes to be inducted, but Hank Aaron obviously was inducted the first year he was eligible.

Going to have to manipulate the hof dataset such that we can append it to each player's hitting totals.

In [15]:
hof.columns

Index(['playerID', 'yearid', 'votedBy', 'ballots', 'needed', 'votes',
       'inducted', 'category', 'needed_note'],
      dtype='object')

In [16]:
hof['category'].value_counts()

Player               3997
Manager                74
Pioneer/Executive      39
Umpire                 10
Name: category, dtype: int64

In [17]:
hof['needed_note'].value_counts()

1st       81
Top 20    76
Name: needed_note, dtype: int64

In [18]:
hof=hof[hof['category'].isin(['Player'])]
hof=hof.drop(columns=['needed_note'])
hof

playerID  yearid votedBy  ballots  needed  votes inducted category
0      cobbty01    1936   BBWAA    226.0   170.0  222.0        Y   Player
1      ruthba01    1936   BBWAA    226.0   170.0  215.0        Y   Player
2     wagneho01    1936   BBWAA    226.0   170.0  215.0        Y   Player
3     mathech01    1936   BBWAA    226.0   170.0  205.0        Y   Player
4     johnswa01    1936   BBWAA    226.0   170.0  189.0        Y   Player
...         ...     ...     ...      ...     ...    ...      ...      ...
4115  grudzma01    2016   BBWAA    440.0   330.0    0.0        N   Player
4116  ausmubr01    2016   BBWAA    440.0   330.0    0.0        N   Player
4117  hamptmi01    2016   BBWAA    440.0   330.0    0.0        N   Player
4118  castilu01    2016   BBWAA    440.0   330.0    0.0        N   Player
4119  lowelmi01    2016   BBWAA    440.0   330.0    0.0        N   Player

[3997 rows x 8 columns]

In [19]:
hof['percent']=hof['votes']/hof['ballots']
hof['threshold']=hof['needed']/hof['ballots']
hof[hof['inducted'].isin(['Y'])]

playerID  yearid votedBy  ballots  needed  votes inducted category  \
0      cobbty01    1936   BBWAA    226.0   170.0  222.0        Y   Player   
1      ruthba01    1936   BBWAA    226.0   170.0  215.0        Y   Player   
2     wagneho01    1936   BBWAA    226.0   170.0  215.0        Y   Player   
3     mathech01    1936   BBWAA    226.0   170.0  205.0        Y   Player   
4     johnswa01    1936   BBWAA    226.0   170.0  189.0        Y   Player   
...         ...     ...     ...      ...     ...    ...      ...      ...   
4055  martipe02    2015   BBWAA    549.0   412.0  500.0        Y   Player   
4056  smoltjo01    2015   BBWAA    549.0   412.0  455.0        Y   Player   
4057  biggicr01    2015   BBWAA    549.0   412.0  454.0        Y   Player   
4088  griffke02    2016   BBWAA    440.0   330.0  437.0        Y   Player   
4089  piazzmi01    2016   BBWAA    440.0   330.0  365.0        Y   Player   

       percent  threshold  
0     0.982301   0.752212  
1     0.951327   0.752212  
2     0.951327   0.752212  
3     0.907080   0.752212  
4     0.836283   0.752212  
...        ...        ...  
4055  0.910747   0.750455  
4056  0.828780   0.750455  
4057  0.826958   0.750455  
4088  0.993182   0.750000  
4089  0.829545   0.750000  

[247 rows x 10 columns]

In [20]:
type(hof['inducted'][0])

str

In [21]:
if hof['inducted'].str.contains('Y').any():
    print('yes')

yes


To get voting data into one column, we're going to show the average % of ballot a player received through the years, and the average % threshold needed.

In [22]:
df=hof[hof['playerID'].isin(['youngro01'])]
if df['inducted'].str.contains('Y').any():
    inducted=df[df['inducted'].isin(['Y'])].reset_index()
    print(inducted['percent'][0])
#     threshold=inducted['threshold'][0]

nan


In [23]:
#how to check if there are any null values in column
#df['your column name'].isnull().values.any()

In [24]:
hof_clean=pd.DataFrame()
for player in hof['playerID'].value_counts().index:
    playerdf=pd.DataFrame()
    
    #extract player data
    df=hof[hof['playerID'].isin([player])].reset_index()
        
    #store information to append
#     induction=df[df['inducted'].isin(['Y'])].reset_index()
    year=df.max()['yearid']
    #voter=df['votedBy'][0]
    
#     df['percent']=df['votes']/df['ballots']
#     df['threshold']=df['needed']/df['ballots']
    
#     ballots=df.sum()['ballots']/len(df)
#     needed=df.sum()['needed']/len(df)
#     votes=df.sum()['votes']/len(df)
    attempts=len(df)
    if df['inducted'].str.contains('Y').any():
        inducted=df[df['inducted'].isin(['Y'])].reset_index()
        percent=inducted['percent'][0]
        threshold=inducted['threshold'][0]
        votedBy=inducted['votedBy'][0]
        playerdf=playerdf.append({'year':year,'player_id':player,'percent':percent,'threshold':
                                threshold,'years':attempts,'votedBy':votedBy,'inducted':'Y'},ignore_index=True)
    else:
#         year=df.max()['yearid']
#         voter=df['votedBy'].value_counts().index[0]
#         ballots=df.sum()['ballots']/len(df)
#         needed=df.sum()['needed']/len(df)
#         votes=df.sum()['votes']/len(df)
#         attempts=len(df)
        max_percent=df.max()['percent']
        threshold=df['threshold'][df['percent'].idxmax()]
        votedBy=df['votedBy'].value_counts().index[0]
        playerdf=playerdf.append({'year':year,'player_id':player,'percent':max_percent,'threshold':
                                threshold,'years':attempts,'votedBy':votedBy,'inducted':'N'},ignore_index=True)
    hof_clean=hof_clean.append(playerdf)

In [25]:
hof_clean=hof_clean.reset_index().drop(columns='index')
hof_clean

inducted   percent  player_id  threshold   votedBy    year  years
0           Y       NaN  roushed01        NaN  Veterans  1962.0   20.0
1           Y       NaN  youngro01        NaN  Veterans  1972.0   19.0
2           Y       NaN  schalra01        NaN  Veterans  1955.0   18.0
3           Y  0.869281  ruffire01        NaN   Run Off  1967.0   18.0
4           Y       NaN  rizzuph01        NaN  Veterans  1994.0   17.0
...       ...       ...        ...        ...       ...     ...    ...
1184        N  0.012821   pikeli01   0.756410  Veterans  1936.0    1.0
1185        N  0.000000  washicl01   0.751064     BBWAA  1996.0    1.0
1186        N  0.000000  coninje01   0.750439     BBWAA  2013.0    1.0
1187        N  0.002114   puhlte01   0.746300     BBWAA  1997.0    1.0
1188        N  0.000000  milleno01   0.750649     BBWAA  1980.0    1.0

[1189 rows x 7 columns]

In [26]:
hof_clean[-hof_clean['votedBy'].isin(['BBWAA'])]['threshold'].isnull().value_counts()

True     140
False     30
Name: threshold, dtype: int64

Of players not voted in by BBWAA, 135/165 did no go through standard voting procedures, or the data was at least not recorded. Let's leave these other voting routes in the dataset, but keep in mind that we may want to take them out.

In [27]:
combined=batting.join(hof_clean.set_index('player_id'), on='player_id')
combined

player_id  stint       g       ab       r       h  double  triple  \
0     mcguide01   37.0  1781.0   6290.0   770.0  1748.0   300.0    79.0   
1     newsobo01   39.0   609.0   1337.0    86.0   253.0    28.0     3.0   
2     henderi01   33.0  3081.0  10961.0  2295.0  3055.0   510.0    66.0   
3      kaatji01   31.0  1004.0   1251.0   117.0   232.0    44.0     5.0   
4      johnto01   30.0   761.0    900.0    43.0   141.0    13.0     0.0   
...         ...    ...     ...      ...     ...     ...     ...     ...   
3798  pasquda01   10.0   905.0   2620.0   341.0   638.0   129.0    15.0   
3799   fainfe01   11.0  1151.0   3930.0   595.0  1139.0   213.0    30.0   
3800  reynoma01   11.0  1258.0   4178.0   612.0   959.0   192.0    13.0   
3801  madsory01   10.0   559.0     48.0     3.0     6.0     1.0     0.0   
3802  brognri01   11.0   848.0   2958.0   379.0   795.0   176.0    13.0   

         hr     rbi  ...   hbp    sh    sf  g_idp  inducted   percent  \
0      45.0   840.0  ...  84.0  43.0   0.0    0.0       NaN       NaN   
1       1.0   102.0  ...   1.0  86.0   0.0   30.0         N  0.094118   
2     297.0  1115.0  ...  98.0  30.0  67.0  172.0         Y  0.948052   
3      16.0   106.0  ...   7.0  34.0   7.0   14.0         N  0.295508   
4       5.0    54.0  ...   1.0  93.0   3.0   20.0         N  0.317254   
...     ...     ...  ...   ...   ...   ...    ...       ...       ...   
3798  117.0   390.0  ...  15.0   9.0  21.0   43.0       NaN       NaN   
3799   48.0   570.0  ...  18.0  44.0   8.0  117.0       NaN       NaN   
3800  237.0   661.0  ...  47.0   3.0  34.0   88.0       NaN       NaN   
3801    0.0     2.0  ...   0.0   7.0   0.0    2.0       NaN       NaN   
3802  106.0   458.0  ...   7.0   5.0  26.0   72.0       NaN       NaN   

      threshold  votedBy    year years  
0           NaN      NaN     NaN   NaN  
1      0.750000    BBWAA  1973.0  13.0  
2      0.751391    BBWAA  2009.0   1.0  
3      0.751773    BBWAA  2003.0  15.0  
4      0.751391    BBWAA  2009.0  15.0  
...         ...      ...     ...   ...  
3798        NaN      NaN     NaN   NaN  
3799        NaN      NaN     NaN   NaN  
3800        NaN      NaN     NaN   NaN  
3801        NaN      NaN     NaN   NaN  
3802        NaN      NaN     NaN   NaN  

[3803 rows x 25 columns]

In [28]:
#take out players that aren't in hof dataset
clean=combined[-combined['inducted'].isnull()]
clean

player_id  stint       g       ab       r       h  double  triple  \
1     newsobo01   39.0   609.0   1337.0    86.0   253.0    28.0     3.0   
2     henderi01   33.0  3081.0  10961.0  2295.0  3055.0   510.0    66.0   
3      kaatji01   31.0  1004.0   1251.0   117.0   232.0    44.0     5.0   
4      johnto01   30.0   761.0    900.0    43.0   141.0    13.0     0.0   
5     carltst01   31.0   745.0   1719.0   123.0   346.0    49.0     6.0   
...         ...    ...     ...      ...     ...     ...     ...     ...   
3729  robinja02   10.0  1382.0   4877.0   947.0  1518.0   273.0    54.0   
3738  torboje01   10.0   574.0   1391.0    78.0   297.0    42.0     3.0   
3745  hornebo01   10.0  1020.0   3777.0   560.0  1047.0   169.0     8.0   
3774  tobinji01   11.0   396.0    796.0    81.0   183.0    35.0     3.0   
3782  kremera01   10.0   309.0    687.0    57.0   122.0    30.0     2.0   

         hr     rbi  ...   hbp     sh    sf  g_idp  inducted   percent  \
1       1.0   102.0  ...   1.0   86.0   0.0   30.0         N  0.094118   
2     297.0  1115.0  ...  98.0   30.0  67.0  172.0         Y  0.948052   
3      16.0   106.0  ...   7.0   34.0   7.0   14.0         N  0.295508   
4       5.0    54.0  ...   1.0   93.0   3.0   20.0         N  0.317254   
5      13.0   140.0  ...  10.0   94.0  13.0   33.0         Y  0.956140   
...     ...     ...  ...   ...    ...   ...    ...       ...       ...   
3729  137.0   734.0  ...  72.0  104.0   9.0  113.0         Y  0.775000   
3738    8.0   101.0  ...   4.0   17.0  10.0   36.0         N  0.000000   
3745  218.0   685.0  ...  16.0    1.0  50.0  111.0         N  0.000000   
3774   17.0   102.0  ...   3.0    5.0   0.0   22.0         N  0.010363   
3782    5.0    69.0  ...   0.0   36.0   0.0    0.0         N  0.008264   

      threshold  votedBy    year years  
1      0.750000    BBWAA  1973.0  13.0  
2      0.751391    BBWAA  2009.0   1.0  
3      0.751773    BBWAA  2003.0  15.0  
4      0.751391    BBWAA  2009.0  15.0  
5      0.750000    BBWAA  1994.0   1.0  
...         ...      ...     ...   ...  
3729   0.750000    BBWAA  1962.0   1.0  
3738   0.750000    BBWAA  1979.0   1.0  
3745   0.750000    BBWAA  1994.0   1.0  
3774   0.751295    BBWAA  1956.0   1.0  
3782   0.752066    BBWAA  1958.0   2.0  

[1125 rows x 25 columns]

In [29]:
clean['inducted'].value_counts()

N    908
Y    217
Name: inducted, dtype: int64

235 players in mlb hof, so this is a pretty good dataset.

In [30]:
clean[clean['votedBy'].isin(['BBWAA'])&clean['inducted'].isin(['Y'])].sort_values(by='h')

player_id  stint       g       ab       r       h  double  triple  \
2242  suttebr01   12.0   661.0    102.0     6.0     9.0     0.0     0.0   
79    gossari01   24.0  1003.0     85.0     1.0     9.0     1.0     0.0   
19    eckerde01   26.0  1073.0    181.0     9.0    24.0     3.0     0.0   
644   fingero01   17.0   945.0    180.0    10.0    31.0     3.0     0.0   
15    wilheho01   31.0  1070.0    432.0    24.0    38.0     3.0     1.0   
...         ...    ...     ...      ...     ...     ...     ...     ...   
67    yastrca01   23.0  3308.0  11988.0  1816.0  3419.0   646.0    59.0   
100   speaktr01   22.0  2789.0  10195.0  1882.0  3514.0   792.0   222.0   
90    musiast01   22.0  3026.0  10972.0  1949.0  3630.0   725.0   177.0   
59    aaronha01   23.0  3298.0  12364.0  2174.0  3771.0   624.0    98.0   
46     cobbty01   24.0  3035.0  11434.0  2246.0  4189.0   724.0   295.0   

         hr     rbi  ...    hbp     sh     sf  g_idp  inducted   percent  \
2242    0.0     6.0  ...    1.0   10.0    1.0    1.0         Y  0.769231   
79      0.0     2.0  ...    0.0    4.0    0.0    2.0         Y  0.858195   
19      3.0    12.0  ...    0.0    5.0    1.0    1.0         Y  0.832016   
644     2.0     9.0  ...    0.0   18.0    0.0    2.0         Y  0.811628   
15      1.0    21.0  ...    0.0   34.0    1.0    8.0         Y  0.837975   
...     ...     ...  ...    ...    ...    ...    ...       ...       ...   
67    452.0  1844.0  ...   40.0   13.0  105.0  323.0         Y  0.946309   
100   117.0  1529.0  ...  103.0  309.0    0.0    0.0         Y  0.820896   
90    475.0  1951.0  ...   53.0   35.0   53.0  243.0         Y  0.932353   
59    755.0  2297.0  ...   32.0   21.0  121.0  328.0         Y  0.978313   
46    117.0  1937.0  ...   94.0  295.0    0.0    0.0         Y  0.982301   

      threshold  votedBy    year years  
2242   0.750000    BBWAA  2006.0  13.0  
79     0.751381    BBWAA  2008.0   9.0  
19     0.750988    BBWAA  2004.0   1.0  
644    0.751163    BBWAA  1992.0   2.0  
15     0.751899    BBWAA  1985.0   8.0  
...         ...      ...     ...   ...  
67     0.751678    BBWAA  1989.0   1.0  
100    0.751244    BBWAA  1937.0   2.0  
90     0.750000    BBWAA  1969.0   1.0  
59     0.751807    BBWAA  1982.0   1.0  
46     0.752212    BBWAA  1936.0   1.0  

[116 rows x 25 columns]

We still have a lot of pitchers in the dataset- whose haitting statistics shouldn't influence the model.

In [31]:
pitchers=pd.read_csv('/kaggle/input/baseball-databank/Pitching.csv')
pitchers

playerID  yearID  stint teamID lgID   W   L   G  GS  CG  ...  IBB  \
0      bechtge01    1871      1    PH1  NaN   1   2   3   3   2  ...  NaN   
1      brainas01    1871      1    WS3  NaN  12  15  30  30  30  ...  NaN   
2      fergubo01    1871      1    NY2  NaN   0   0   1   0   0  ...  NaN   
3      fishech01    1871      1    RC1  NaN   4  16  24  24  22  ...  NaN   
4      fleetfr01    1871      1    NY2  NaN   0   1   1   1   1  ...  NaN   
...          ...     ...    ...    ...  ...  ..  ..  ..  ..  ..  ...  ...   
44134  youngch03    2015      1    KCA   AL  11   6  34  18   0  ...  0.0   
44135  zieglbr01    2015      1    ARI   NL   0   3  66   0   0  ...  3.0   
44136  zimmejo02    2015      1    WAS   NL  13  10  33  33   0  ...  3.0   
44137   zitoba01    2015      1    OAK   AL   0   0   3   2   0  ...  0.0   
44138   zychto01    2015      1    SEA   AL   0   0  13   1   0  ...  0.0   

        WP  HBP  BK    BFP    GF    R   SH   SF  GIDP  
0      NaN  NaN   0    NaN   NaN   42  NaN  NaN   NaN  
1      NaN  NaN   0    NaN   NaN  292  NaN  NaN   NaN  
2      NaN  NaN   0    NaN   NaN    9  NaN  NaN   NaN  
3      NaN  NaN   0    NaN   NaN  257  NaN  NaN   NaN  
4      NaN  NaN   0    NaN   NaN   21  NaN  NaN   NaN  
...    ...  ...  ..    ...   ...  ...  ...  ...   ...  
44134  5.0  0.0   0  500.0   3.0   44  4.0  2.0   NaN  
44135  2.0  1.0   0  263.0  46.0   17  1.0  0.0   NaN  
44136  2.0  8.0   1  831.0   0.0   89  8.0  2.0   NaN  
44137  0.0  0.0   0   37.0   1.0    8  0.0  0.0   NaN  
44138  1.0  2.0   0   76.0   4.0    6  0.0  0.0   NaN  

[44139 rows x 30 columns]

A lot of players in the old days, and even today, will pitch even if they're a position player. Thus, they show up on the pithers list even though they're hall of famers because of their hitting. Account for this by cutting all with less than 100 hits.

In [32]:
pitchers_list=[]
for player in clean['player_id']:
    if pitchers['playerID'].str.contains(player).any():
        df=clean[clean['player_id'].isin([player])].reset_index()
        
        #weed out hitters that still show up, if they have less than 1000 hits then they are not in the
        #hof because of their hitting statistics
        if df['h'][0] <1000:
            pitchers_list.append(player)

In [33]:
pitchers_list

['newsobo01',
 'kaatji01',
 'johnto01',
 'carltst01',
 'ryanno01',
 'niekrph01',
 'oroscje01',
 'houghch01',
 'mulhote01',
 'wilheho01',
 'eckerde01',
 'reussje01',
 'maddugr01',
 'suttodo01',
 'perryga01',
 'darwida01',
 'niekrjo01',
 'hoytwa01',
 'zachato01',
 'clemero02',
 'martide01',
 'blylebe01',
 'quinnja01',
 'honeyri01',
 'wellsda01',
 'johnsra05',
 'tananfr01',
 'wynnea01',
 'candejo01',
 'alexado01',
 'grimebu01',
 'pennohe01',
 'millebo04',
 'gossari01',
 'simmocu01',
 'youngcy01',
 'ruffire01',
 'mcmahdo02',
 'glavito02',
 'pizarju01',
 'sadecra01',
 'smoltjo01',
 'jonessa01',
 'spahnwa01',
 'mesajo01',
 'griffcl01',
 'mcdanli01',
 'seaveto01',
 'gordoto01',
 'leiteal01',
 'smithle02',
 'reuscri01',
 'jacksgr01',
 'altroni01',
 'alexape01',
 'roberro01',
 'lyonste01',
 'francjo01',
 'johnswa01',
 'waltebu01',
 'burdele01',
 'schilcu01',
 'rixeyep01',
 'timlimi01',
 'torremi01',
 'rogerke01',
 'hardeme01',
 'luquedo01',
 'faberre01',
 'koosmje01',
 'jenkife01',
 'frymawo01'

In [34]:
hitters=clean[-clean['player_id'].isin(pitchers_list)]
hitters

player_id  stint       g       ab       r       h  double  triple  \
2     henderi01   33.0  3081.0  10961.0  2295.0  3055.0   510.0    66.0   
8     baineha01   32.0  2830.0   9908.0  1299.0  2866.0   488.0    49.0   
9     ansonca01   27.0  2523.0  10277.0  1996.0  3418.0   581.0   142.0   
16    dempsri01   26.0  1766.0   4692.0   525.0  1093.0   223.0    12.0   
18    bucknbi01   28.0  2517.0   9397.0  1077.0  2715.0   498.0    49.0   
...         ...    ...     ...      ...     ...     ...     ...     ...   
3664  camparo01   10.0  1215.0   4205.0   627.0  1161.0   178.0    18.0   
3693   remyje01   10.0  1154.0   4455.0   605.0  1226.0   140.0    38.0   
3729  robinja02   10.0  1382.0   4877.0   947.0  1518.0   273.0    54.0   
3738  torboje01   10.0   574.0   1391.0    78.0   297.0    42.0     3.0   
3745  hornebo01   10.0  1020.0   3777.0   560.0  1047.0   169.0     8.0   

         hr     rbi  ...   hbp     sh    sf  g_idp  inducted   percent  \
2     297.0  1115.0  ...  98.0   30.0  67.0  172.0         Y  0.948052   
8     384.0  1628.0  ...  14.0    9.0  99.0  298.0         N  0.061224   
9      97.0  2076.0  ...  32.0   27.0   0.0    0.0         Y       NaN   
16     96.0   471.0  ...  18.0   63.0  42.0  131.0         N  0.002114   
18    174.0  1208.0  ...  42.0   47.0  97.0  247.0         N  0.021277   
...     ...     ...  ...   ...    ...   ...    ...       ...       ...   
3664  242.0   856.0  ...  30.0   30.0  18.0  143.0         Y  0.794118   
3693    7.0   329.0  ...   4.0  116.0  32.0   83.0         N  0.002252   
3729  137.0   734.0  ...  72.0  104.0   9.0  113.0         Y  0.775000   
3738    8.0   101.0  ...   4.0   17.0  10.0   36.0         N  0.000000   
3745  218.0   685.0  ...  16.0    1.0  50.0  111.0         N  0.000000   

      threshold     votedBy    year years  
2      0.751391       BBWAA  2009.0   1.0  
8      0.751391       BBWAA  2011.0   5.0  
9           NaN  Old Timers  1939.0   2.0  
16     0.750529       BBWAA  1998.0   1.0  
18     0.751064       BBWAA  1996.0   1.0  
...         ...         ...     ...   ...  
3664   0.750000       BBWAA  1969.0   7.0  
3693   0.750000       BBWAA  1990.0   1.0  
3729   0.750000       BBWAA  1962.0   1.0  
3738   0.750000       BBWAA  1979.0   1.0  
3745   0.750000       BBWAA  1994.0   1.0  

[740 rows x 25 columns]

In [35]:
hitters['inducted'].value_counts()

N    590
Y    150
Name: inducted, dtype: int64

67 HOFers taken out, this seems accurate in terms of getting rid of pitchers.

In [36]:
hitters.sort_values(by='hr',ascending=False).head(10)

player_id  stint       g       ab       r       h  double  triple     hr  \
117  bondsba01   22.0  2986.0   9847.0  2227.0  2935.0   601.0    77.0  762.0   
59   aaronha01   23.0  3298.0  12364.0  2174.0  3771.0   624.0    98.0  755.0   
92    ruthba01   22.0  2503.0   8398.0  2174.0  2873.0   506.0   136.0  714.0   
55    mayswi01   24.0  2992.0  10881.0  2062.0  3283.0   523.0   140.0  660.0   
65   griffke02   24.0  2671.0   9801.0  1662.0  2781.0   524.0    38.0  630.0   
397   sosasa01   20.0  2354.0   8813.0  1475.0  2408.0   379.0    45.0  609.0   
101  robinfr02   23.0  2808.0  10006.0  1829.0  2943.0   528.0    72.0  586.0   
669  mcgwima01   18.0  1874.0   6187.0  1167.0  1626.0   252.0     6.0  583.0   
102  killeha01   22.0  2435.0   8147.0  1283.0  2086.0   290.0    24.0  573.0   
275  palmera01   20.0  2831.0  10472.0  1663.0  3020.0   585.0    38.0  569.0   

        rbi  ...    hbp     sh     sf  g_idp  inducted   percent  threshold  \
117  1996.0  ...  106.0    4.0   91.0  165.0         N  0.443182   0.750000   
59   2297.0  ...   32.0   21.0  121.0  328.0         Y  0.978313   0.751807   
92   2217.0  ...   43.0  113.0    0.0    2.0         Y  0.951327   0.752212   
55   1903.0  ...   44.0   13.0   91.0  251.0         Y  0.946759   0.750000   
65   1836.0  ...   81.0    8.0  102.0  199.0         Y  0.993182   0.750000   
397  1667.0  ...   59.0   17.0   78.0  202.0         N  0.124780   0.750439   
101  1812.0  ...  198.0   17.0  102.0  269.0         Y  0.891566   0.751807   
669  1414.0  ...   75.0    3.0   78.0  147.0         N  0.237477   0.751391   
102  1584.0  ...   48.0    0.0   77.0  243.0         Y  0.831266   0.751861   
275  1835.0  ...   87.0   15.0  119.0  231.0         N  0.125654   0.750436   

     votedBy    year years  
117    BBWAA  2016.0   4.0  
59     BBWAA  1982.0   1.0  
92     BBWAA  1936.0   1.0  
55     BBWAA  1979.0   1.0  
65     BBWAA  2016.0   1.0  
397    BBWAA  2016.0   4.0  
101    BBWAA  1982.0   1.0  
669    BBWAA  2016.0  10.0  
102    BBWAA  1984.0   4.0  
275    BBWAA  2014.0   4.0  

[10 rows x 25 columns]

Top 10 home run hitters is correct- just making sure I'm not missing a significant amount of data. Notice that players like Alex Rodriguez and Albert Pujols are not on here- this is because as of 2016 they were still active players or were not yet eligible for the hall of fame.

In [37]:
#reset indices
hitters=hitters.reset_index().drop(columns='index')

In [38]:
hitters.to_csv('hitters2.csv')